core/class/jMQTT.class.php

```diff
@@ -130,6 +141,83 @@ public static function templateParameters($_template = ''){
 		return $return;
 	}
 
+	/**
+	 * Return a list of all templates name and file.
+	 * @return list of name and file array.
+	 */
+	public static function templateList(){
+		// log::add('jMQTT', 'debug', 'templateList()');
+		$return = array();
+		// Get personal templates
+		foreach (ls(dirname(__FILE__) . '/../../data/template', '*.json', false, array('files', 'quiet')) as $file) {
+			try {
+				$content = file_get_contents(dirname(__FILE__) . '/../../data/template/' . $file);
+				if (is_json($content)) {
+					foreach (json_decode($content, true) as $k => $v)
+						$return[] = array('[Perso] '.$k, 'plugins/jMQTT/data/template/' . $file);
+				}
+			} catch (Throwable $e) {}
+		}
+		// Get official templates
+		foreach (ls(dirname(__FILE__) . '/../config/template', '*.json', false, array('files', 'quiet')) as $file) {
+			try {
+				$content = file_get_contents(dirname(__FILE__) . '/../config/template/' . $file);
+				if (is_json($content)) {
+					foreach (json_decode($content, true) as $k => $v)
+						$return[] = array($k, 'plugins/jMQTT/core/config/template/' . $file);
+				}
+			} catch (Throwable $e) {}
+		}
+		return $return;
+	}
+
+	/**
+	 * Return one templates content (from json file name).
+	 * @param string $_filename template name to look for
+	 * @return array
+	 */
+	public static function templateByFile($_filename = ''){
+		// log::add('jMQTT', 'debug', 'templateByFile("' . $_filename . '")');
+		$existing_files = jMQTT::templateList();
+		$exists = false;
+		foreach ($existing_files as list($n, $f))
+			if ($f == $_filename) {
+				$exists = true;
+				break;
+			}
+		if (!$exists)
+			throw new Exception(__('Le template demandé n\'existe pas !', __FILE__));
+		// log::add('jMQTT', 'debug', '    get='.dirname(__FILE__) . '/../../../../' . $_filename);
+		try {
+			$content = file_get_contents(dirname(__FILE__) . '/../../../../' . $_filename);
+			if (is_json($content)) {
+				foreach (json_decode($content, true) as $k => $v)
+					return $v;
+			}
+		} catch (Throwable $e) {}
+		return array();
+	}
+
+	/**
+	 * Deletes user defined template by filename.
+	 * @param string $_template template name to look for.
+	 */
+	public static function deleteTemplateByFile($_filename){
+		// log::add('jMQTT', 'debug', 'deleteTemplateByFile("' . $_filename . '")');
+		if (!isset($_filename) || is_null($_filename) || $_filename == '')
+			return false;
+		$existing_files = jMQTT::templateList();
+		$exists = false;
+		foreach ($existing_files as list($n, $f))
+			if ($f == $_filename) {
+				$exists = true;
+				break;
+			}
+		if (!$exists)
+			return false;
+		return unlink(dirname(__FILE__) . '/../../../../' . $_filename);
+	}
+
```

Comment:
Warning, all Jeedom SA plugins using template always refer to name in the json and never to the filename.
It is slower because it parses all template files one by one to the one with the right name inside but ...
... there should be a reason ... maybe ?

Reply:
Yes, that's the way I did it first (see https://github.com/BadWolf42/jMQTT/blame/c29b0ad54af96b2823fd886e92309b7d7aa22857/core/class/jMQTT.class.php#L178-L201)
But, when importing a template file with existing template name, I got multiple times the same template name in the list and could not easily delete just a specific one...
So I implemented a file name based delete and planned to use filename instead of template name all over jMQTT.

I don't see any advantage to use template name and don't understand why I should be preferred.